compiler/rustc_interface/src/passes.rs

```diff
@@ -510,6 +575,18 @@ fn write_out_deps(tcx: TyCtxt<'_>, outputs: &OutputFilenames, out_filenames: &[P
                 }
             }
 
+            // If caller requested this information, add special comments about source file checksums.
+            // These are not necessarily the same checksums as was used in the debug files.
+            if sess.opts.unstable_opts.checksum_hash_algorithm().is_some() {
+                for (path, file_len, checksum_hash) in
+                    files.iter().filter_map(|(path, file_len, hash_algo)| {
+                        hash_algo.map(|hash_algo| (path, file_len, hash_algo))
+                    })
+                {
+                    writeln!(file, "# checksum:{checksum_hash} file_len:{file_len} {path}")?;
```

Comment:
Is it possible that we do this in a single pass like https://github.com/rust-lang/cargo/pull/14137#discussion_r1767916886 proposed?

```
src/lib.rs: src/someotherfile.rs # checksum:blake3=660e91cbd3 file_len:9
```

Reply:
Honestly I don't even know if it is a valid Makefile syntax 😅.

Reply:
So the dependency relationship here is one to many instead of one to one. So I tweaked your proposal slightly, but it is implemented in f43d31097cfca0a86c921ffdd9f2a22241653b25

Reply:
Sorry I didn't mean to rush you into making a change. since you have pointed out that the dependency relationship is not one to one, f43d31097cfca0a86c921ffdd9f2a22241653b25 doesn't seem better to me. I was wrong. Like, if we have multiple different files emitted, those information will be duplicate a lot:

```
/foo/target/debug/deps/libfoo-e93fc74c04a90e85.rmeta: src/main.rs src/a.rs # checksum:blake3=26aa07e1adab
787246f9d333be65d2eb78dd5fd0fee834ba7a769098b4b651bc file_len:19 checksum:blake3=af1349b9f5f9a1a6a0404dea36dcc9499bcb25c9adc112b7cc9a93cae41f326
2 file_len:0

/foo/target/debug/deps/foo-e93fc74c04a90e85.d: src/main.rs src/a.rs # checksum:blake3=26aa07e1adab787246f
9d333be65d2eb78dd5fd0fee834ba7a769098b4b651bc file_len:19 checksum:blake3=af1349b9f5f9a1a6a0404dea36dcc9499bcb25c9adc112b7cc9a93cae41f3262 file_
len:0
```

What do you think?

Reply:
You're right, and I should have pushed back on this suggestion as well. I'm not sure what I thought this was going to accomplish in retrospect. At this rate I think we'll need to address it in a follow-up PR. This PR is already being rolled up.